### PR TITLE
[Refactor:System] Stores Groups in Master DB

### PR DESCRIPTION
**Pull request by cjreed121**
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?
Right now when creating a new course through the website, you must pick a course to base from. This simply finds a group to give the new course.

### What is the new behavior?
Rather than ask for a course, the site just asks for the group directly which comes from a new table. This new table will also be used in plagiarism.

### CODE REVIEW

File: migration/migrator/data/submitty_db.sql [hunk 1]
```diff
@@ -298,6 +298,17 @@ CREATE TABLE public.courses (
...
 
+--
+-- Name: courses_groups; Type: TABLE; Schema: public; Owner: -
+--
+
+CREATE TABLE public.courses_groups (
+    semester character varying(255) NOT NULL,
+    course character varying(255) NOT NULL,
+    group_name character varying(255) NOT NULL
+);
```
Comment (MasterOdin) [inline migration/migrator/data/submitty_db.sql:309]: Why have this be a separate table from `courses`?

**Review by sbelsk — COMMENTED** [review]
For the plagiarism backend, it would be convenient if we had a function that takes a semester parameter and a course id parameter, and returns the group that corresponds with that course

File: migration/migrator/migrations/master/20210714135438_add_course_groups.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    :type database: migrator.db.Database
+    """
+    database.execute("ALTER TABLE courses ADD COLUMN IF NOT EXISTS group_name varchar(255);")
+    database.execute("ALTER TABLE courses ADD COLUMN IF NOT EXISTS owner_name varchar(255);")
+    database.execute("ALTER TABLE courses DROP CONSTRAINT IF EXISTS group_validate;")
+    database.execute("ALTER TABLE courses DROP CONSTRAINT IF EXISTS owner_validate;")
+    database.execute("ALTER TABLE courses ADD CONSTRAINT group_validate CHECK (group_name ~ '^[a-zA-Z0-9_-]*$')")
+    database.execute("ALTER TABLE courses ADD CONSTRAINT owner_validate CHECK (owner_name ~ '^[a-zA-Z0-9_-]*$')")
+    semester_dir = Path(config.submitty['submitty_data_dir'], 'courses')
+    for semester in semester_dir.iterdir():
```
Comment (bmcutler) [inline migration/migrator/migrations/master/20210714135438_add_course_groups.py:22]: This isn't safe to do on production.  We have directories from old terms that do not have corresponding Submitty databases / are not in the master database.

So we'll need to wrap this with some checks/exception handling.

File: migration/migrator/migrations/master/20210714135438_add_course_groups.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            query = "UPDATE courses SET group_name=:g, owner_name=:o WHERE semester = :s AND course = :c;"
+            params = {
+                's': semester.name,
+                'c': course.name,
+                'g': course.group(),
+                'o': course.owner()
+            }
+            database.session.execute(text(query), params)
+    # we should be able to force the not null constraint now
+    database.execute("ALTER TABLE courses ALTER COLUMN group_name SET NOT NULL;")
```
Comment (bmcutler) [inline migration/migrator/migrations/master/20210714135438_add_course_groups.py:33]: And perhaps the reverse is true, a course in the database that is missing a directory (if old course directories are not currently mounted on the file system or are inaccessible).  

Let's write this migration more defensively.  We can set the user and group to root if the directory does not exist or if the directory is inaccessible.

File: site/app/libraries/database/DatabaseQueries.php [hunk 4]
```diff
@@ -7355,6 +7355,15 @@ public function getGradeableAccessTeam(
         return $this->course_db->rows();
     }
 
+    public function getUserGroups(string $user_id): array {
```
Comment (bmcutler) [inline site/app/libraries/database/DatabaseQueries.php:7358]: What is the purpose of this function?  It should only grab the group for courses in which that user has instructor access?

**Review by bmcutler — CHANGES_REQUESTED** [review]
Just read through the code, I would still like to test it after these comments are addressed.

File: migration/migrator/migrations/master/20210714135438_add_course_groups.py (hunk 2 above)
Comment (cjreed121) [reply migration/migrator/migrations/master/20210714135438_add_course_groups.py:22]: The UPDATE query is actually not dangerous or unsafe. If it does not find a course in the courses table then the query does nothing and moves on without any errors.

File: migration/migrator/migrations/master/20210714135438_add_course_groups.py (hunk 3 above)
Comment (cjreed121) [reply migration/migrator/migrations/master/20210714135438_add_course_groups.py:35]: This is now properly set before we enforce the not null constraint.

File: site/app/libraries/database/DatabaseQueries.php (hunk 4 above)
Comment (cjreed121) [reply site/app/libraries/database/DatabaseQueries.php:7358]: I added a check to make sure the user is an instructor of that course. The purpose of this function is to feed to the front end for faculty to choose the group they want to use.

**Review by bmcutler — APPROVED** [review]

**Pull request merged**